### ISSUE TITLE
[TODO] Import images

### ISSUE THREAD

**Issue by vjeux** [vjeux/excalibur#19]
This would be nice, but may be annoying to implement persistence with it

**Comment by jaredpalmer** [comment]
How about in the interim we support insertion from URL?

**Comment by jaredpalmer** [comment]
From UX perspective, you should be able to:

- Drag and drop any image onto the window/canvas
- Paste from image clipboard
- Change fill mode: Fill, Contain, Tile

**Comment by cmart11** [comment]
Seems like we should also be able to crop the image with a shape or upload it as a shape's fill, but that may be better left for a new issue.

**Comment by jordic** [comment]
Persistence could be done as base64 encoded json values

**Comment by dwelle** [comment]
Keep in mind that we still store to `localStorage` (so we have ~5MB limit) and need to cache app (+canvas) state (potentially multiple, when #253 is implemented). Copy/pasting (meaning potentially another whole canvas) logic will also use `localStorage`.

Maybe we should investigate using different storage (IndexedDB...), soon.

**Comment by jordic** [comment]
Perhaps the implementations should be every image blob a key on localstorage and only store on the file the reference to the item. Others are doing these, (thought kibana)

**Comment by j-f1** [comment]
We could upload the images in an encrypted manner to `json.excalidraw.com`.

**Comment by lipis** [comment]
What about downloading the file?

**Comment by j-f1** [comment]
Download a zip maybe?

**Comment by garyelephant** [comment]
+1

**Comment by kbariotis** [comment]
How about getting a non-profit license from a hosting service? I've used [Cloudinary](https://cloudinary.com/solutions/industries/cloudinary-for-nonprofits) in the past. cc @vjeux

**Comment by kbariotis** [comment]
How about getting a non-profit license from a hosting service? I've used [Cloudinary](https://cloudinary.com/solutions/industries/cloudinary-for-nonprofits) in the past. cc @vjeux

**Comment by lipis** [comment]
What's wrong with uploading them to json.excalidraw.com? (If we are going to that route)

**Comment by vjeux** [comment]
Yeah, I’d rather not we use random third party to host user content. We already have a way to store end to end encrypted blobs with json.excalidraw.com

**Comment by moritonal** [comment]
I've worked on this for a couple of hours and found it fairly simple to add a new component so that's kudos to neat code. I've noticed there's PR #1589 which is also implementing this, so will pivot to help that progress.

**Comment by chendabo** [comment]
is there any progress on this issue?

**Comment by dwelle** [comment]
Not at this moment.

I'm gonna link here what the implementation should look like: https://github.com/excalidraw/excalidraw/pull/1589#issuecomment-716223677

**Comment by o0O-ds-O0o** [comment]
Hi community of excalidraw!   =)

I’m Daniel and i’m working as an assistant at the computer center of Bauhaus-Universität Weimar. My job is to research on open source alternatives fitting to the needs of the people of our university. Further I am doing my master thesis „Sources of Freedom“ about the potential of FLOSS for artists and designers.

There is a professor for industrial design, Andreas Mühlenberend, who understands the benefit of community driven open source projects and he likes excalidraw very much! Especially the wiggling lines
A feature to import images would be killer for the use of excalidraw in his every day lessons. This professor asked me to ask you, if it would be possible to pay someone from your community to implement it. And if so, how much about it would cost?

With best regards from Weimar, Germany,
Daniel

**Comment by Kartik1397** [comment]
Can I work on this feature?

**Comment by lipis** [comment]
So it is decided? We are going the cloud way? How it's going to work when saving the file? Do you need internet to see it? How big much can we get from local storage?

**Comment by dwelle** [comment]
The cloud will only be used for when persisting collaboration scenes to server (technically, link-shared scenes could do that as well). When saving the file locally, the image will be encoded into the JSON file in base64.

Though it seems we haven't specced that part yet :).

But you raise a good question about browser-persisted scenes (LS). I guess we'll need to start using IDB after all. Alright, this is getting more complex by the minute.

The sad part is, all of this must be implemented together in the first take because we can't support images in a way that will error out when you start collaboration session, or refresh browser, or save the file locally. It's either all or nothing.

**Comment by lipis** [comment]
Let's write an RFC in another issue with all the technical stuff that we need as a check list and modify that one.

**Comment by heitara** [comment]
I've create the following [PR #3424](https://github.com/excalidraw/excalidraw/pull/3424). It covers just the basics, but I can try to finish the implementation. 
@lipis could you please try to summarize what else should be covered? 
@dwelle I have some experience with Firebase, so if you have a clear idea what we want I can try to implement it.

**Comment by SheetalPratik09** [comment]
I am still confused. Can I add logos or pictures. It is preventing me from creating effective diagram.

**Comment by heitara** [comment]
This is in development. It won't be released soon. You can use the excalidraw [library](https://libraries.excalidraw.com/?target=_excalidraw&referrer=https%3A%2F%2Fexcalidraw.com%2F&useHash=true&token=GPIgs4wTFJpgGbk2WAsbk&sort=default) to add some logos or symbols.

**Comment by theoryshaw** [comment]
+1

**Comment by DuncanLithgow** [comment]
@o0O-ds-O0o note that anyone can support the project at any time from https://opencollective.com/excalidraw

**Comment by jaideeprana** [comment]
Any tentative date on when this feature will be released?

**Comment by heitara** [comment]
Check the PR mentioned above.

**Comment by dragos-vlad** [comment]
Any updates on this issue? The PR referenced is a POC and hasn't had any activity in 2021.

**Comment by dwelle** [comment]
Wrong PR. This one is active: https://github.com/excalidraw/excalidraw/pull/3424. Will get back to it in a few days. Latest preview: https://excalidraw-9im2ruy53-excalidraw.vercel.app/

**Comment by bric3** [comment]
While this PR is being worked on. For my [_IntelliJ IDEA Excalidraw integration plugin_](https://plugins.jetbrains.com/plugin/17096-excalidraw-integration/) I copied _internal_ code to support image export / import as suggested by  (although I am currently using version 0.8).

Eg (along with a few other files, like `codec.ts`) :

https://github.com/bric3/excalidraw-jetbrains-plugin/blob/main/excalidraw-assets/src/image.ts

This is definitely a workaround but it's working.

**Comment by ad1992** [comment]
We need to finalise few API's so probably by end next month we will be shipping the next release

**Comment by gohma231** [comment]
Would it be possible for this to support PDFs as well?

**Comment by heitara** [comment]
@gohma231  what do you mean? If you convert the pdf file to an image then you can use it. If you have something else in mind, please create an issue and elaborate on it.

**Comment by ricky11** [comment]
> Seems like we should also be able to crop the image with a shape or upload it as a shape's fill, but that may be better left for a new issue.

would love to be able to crop images.

**Comment by wutsminame** [comment]
> @gohma231 what do you mean? If you convert the pdf file to an image then you can use it. If you have something else in mind, please create an issue and elaborate on it.

is it possible to load the pdf file in excalidraw so that i can read and annotate texts or pictures in pdf via excalidraw and automatically (or manually) generate elements in canvas of excalidraw? what inspired me is an app called guga compatible only with ipad currently, and since that app is similar to excalidraw in terms of a whiteboard, excalidraw can work better with obsidian if some functions(especially with pdf) can be realized.

**Comment by anatoliibakhtin** [comment]
@wutsminame you can: 
https://www.youtube.com/watch?v=nB4cOfn0xAs&ab_channel=Zsolt%27sVisualPersonalKnowledgeManagement
(maybe you already know)
